packages/go/analysis/ad/esc4.go

```diff
@@ -0,0 +1,208 @@
+	"github.com/specterops/bloodhound/dawgs/ops"
+	"github.com/specterops/bloodhound/dawgs/query"
+	"github.com/specterops/bloodhound/dawgs/util/channels"
+	"github.com/specterops/bloodhound/graphschema/ad"
+	"github.com/specterops/bloodhound/log"
+)
+
+func PostADCSESC4(ctx context.Context, tx graph.Transaction, outC chan<- analysis.CreatePostRelationshipJob, groupExpansions impact.PathAggregator, enterpriseCA, domain *graph.Node, cache ADCSCache) error {
+	// 1.
+	principals := cardinality.NewBitmap32()
+
+	// 2. iterate certtemplates that have an outbound `PublishedTo` edge to eca
+	for _, certTemplate := range cache.PublishedTemplateCache[enterpriseCA.ID] {
+		if principalsWithGenericWrite, err := FetchPrincipalsWithGenericWriteOnCertTemplate(tx, certTemplate); err != nil {
+			log.Warnf("error fetching principals with %s on cert template: %v", ad.GenericWrite, err)
+		} else if principalsWithEnrollOrAllExtendedRights, err := FetchPrincipalsWithEnrollOrAllExtendedRightsOnCertTemplate(tx, certTemplate); err != nil {
+			log.Warnf("error fetching principals with %s or %s on cert template: %v", ad.Enroll, ad.AllExtendedRights, err)
+		} else if principalsWithPKINameFlag, err := FetchPrincipalsWithWritePKINameFlagOnCertTemplate(tx, certTemplate); err != nil {
+			log.Warnf("error fetching principals with %s on cert template: %v", ad.WritePKINameFlag, err)
+		} else if principalsWithPKIEnrollmentFlag, err := FetchPrincipalsWithWritePKIEnrollmentFlagOnCertTemplate(tx, certTemplate); err != nil {
+			log.Warnf("error fetching principals with %s on cert template: %v", ad.WritePKIEnrollmentFlag, err)
+		} else if enrolleeSuppliesSubject, err := certTemplate.Properties.Get(string(ad.EnrolleeSuppliesSubject)).Bool(); err != nil {
+			log.Warnf("error fetching %s property on cert template: %v", ad.EnrolleeSuppliesSubject, err)
+		} else if requiresManagerApproval, err := certTemplate.Properties.Get(string(ad.RequiresManagerApproval)).Bool(); err != nil {
+			log.Warnf("error fetching %s property on cert template: %v", ad.RequiresManagerApproval, err)
+		} else {
+
+			// 2a. principals that control the cert template
+			principals.Or(
+				CalculateCrossProductNodeSets(
+					groupExpansions,
+					cache.EnterpriseCAEnrollers[enterpriseCA.ID],
+					cache.CertTemplateControllers[certTemplate.ID],
+				))
+
+			// 2b. principals with `Enroll/AllExtendedRights` + `Generic Write` combination on the cert template
+			principals.Or(
+				CalculateCrossProductNodeSets(
+					groupExpansions,
+					cache.EnterpriseCAEnrollers[enterpriseCA.ID],
+					principalsWithGenericWrite.Slice(),
+					principalsWithEnrollOrAllExtendedRights.Slice(),
+				),
+			)
+
+			// 2c. kick out early if cert template does meet conditions for ESC4
+			if valid, err := isCertTemplateValidForESC4(certTemplate); err != nil {
+				log.Warnf("error validating cert template %d: %v", certTemplate.ID, err)
+				continue
+			} else if !valid {
+				continue
+			}
```

Comment:
Moving this to the start of the if/else chain after line 38 would save us from querying the db unnecessarily in case the certTemplate is not valid.

Reply:
Oh actually this looks to follow the order that was specified in the design doc so it tracks even though it seems a bit different than the logic for other escalations.